### ISSUE TITLE
QM7 + QM7b benchmark hyperparameters

### ISSUE THREAD

**Issue by joegomes** [deepchem/deepchem#415]
The TF model hyperparameters used for QM7 and QM7b in `benchmark.py` should be updated to match those in `examples/qm7`.  Here's the model:

model = dc.models.TensorflowMultiTaskFitTransformRegressor(
    n_tasks=1, n_features=[23, 23], learning_rate=0.001 , momentum=.8, batch_size=25,
    weight_init_stddevs=[1/np.sqrt(400),1/np.sqrt(100),1/np.sqrt(100)],
    bias_init_consts=[0.,0.,0.], layer_sizes=[400,100,100], 
    dropouts=[0.01,0.01,0.01], fit_transformers=fit_transformers, n_evals=10, seed=123)

and results:

computed_metrics: [9.4825899911757379]
computed_metrics: [0.99697709686330227]
Train scores [kcal/mol]
{'pearson_r2_score': 0.99697709686330227, 'mean_absolute_error': 9.4825899911757379}
computed_metrics: [10.702815291285313]
computed_metrics: [0.99575343313015841]
Valid scores [kcal/mol]
{'pearson_r2_score': 0.99575343313015841, 'mean_absolute_error': 10.702815291285313}
computed_metrics: [10.186901162009551]
computed_metrics: [0.99632169203016496]
Test scores [kcal/mol]
{'pearson_r2_score': 0.99632169203016496, 'mean_absolute_error': 10.186901162009551}

Also, the results improve to about 4-5 kcal/mol MAE if `fit()` is run for 200 epochs instead of 50.

**Comment by miaecle** [comment]
@joegomes Thanks a lot for tweeking the hyperparameters here, going to run on these settings again and update the performance chart. BTW, is it useful if we list the mae for all the tasks in qm7b?

**Comment by joegomes** [comment]
@miaecle I think for now average mae is fine, I haven't done any optimization for the multitask tf example yet (forthcoming PR).  Thanks for updating the example and performance chart!

**Comment by vid33** [comment]
Hi all,
I've been trying to reproduce qm7 GraphConv singletask results reported in the MoleculeNet problem, and after trying a number of things, including different splits, am getting nowhere close. Most R^2 values are > 0.9, and I'm not getting anything much above 0.7. I've attached the top level script I'm using. Any thoughts would be much appreciated.

%%%%%%%%%%%%%%%%%%

"""
Script that trains GraphConv models on qm7 dataset.
"""
from __future__ import print_function
from __future__ import division
from __future__ import unicode_literals

import numpy as np
np.random.seed(123)
import tensorflow as tf
tf.set_random_seed(123)
import deepchem as dc

# Load QM7 dataset
tasks, datasets, transformers = dc.molnet.load_qm7_from_mat(
    featurizer='GraphConv', split='index')
#tasks, datasets, transformers = dc.molnet.load_qm7_from_mat(
#    featurizer='GraphConv', split='random')
train_dataset, valid_dataset, test_dataset = datasets

# Fit models
#metric = [
#	dc.metrics.Metric(dc.metrics.pearson_r2_score, mode="regression"),
#    dc.metrics.Metric(dc.metrics.mean_absolute_error, mode="regression")
#]
metric = [dc.metrics.Metric(dc.metrics.pearson_r2_score, np.mean)]

# Batch size of models
batch_size = 50

model = dc.models.GraphConvTensorGraph(
    				len(tasks), 
    				model_dir = "qm7_tg_GraphConv_data/weights",
    				batch_size=batch_size, 
    				learning_rate=5e-4, 
    				mode="regression")

# Restart from a previous state
#model.global_step = 8600
#model.last_checkpoint = 'qm7_tg_GraphConv_data/weights/model-' + str(model.global_step)

# Fit trained model
model.fit(train_dataset, nb_epoch=100, checkpoint_interval=100, max_checkpoints_to_keep=200)

print("Evaluating model")
train_scores = model.evaluate(train_dataset, metric, transformers)
valid_scores = model.evaluate(valid_dataset, metric, transformers)

print("Train scores")
print(train_scores)

print("Validation scores")
print(valid_scores)

**Comment by rbharath** [comment]
@vid33 It looks like you're using default hyperparams on `GraphConvTensorGraph`. Try using the hyperparams from the paper added to S3 in #880

**Comment by vid33** [comment]
I was trying to reproduce the qm7 results. As far as I can tell the appropriate hyperparameters for e.g. GraphConv on S3 are ('qm7graphconvreg.pkl' file):

{'batch_size': 241,
 'learning_rate': 0.0005683155578218224,
 'n_filters': 202,
 'n_fully_connected_nodes': 293,
 'nb_epoch': 20,
 'seed': 123}

At the level of the network & the info in the MoleculeNet paper, I understand this to mean: Two GraphConv layers, with 202 features at outputs, both followed by a batch norm and GraphPool layer, followed by a dense layer with 293 outputs, followed by GraphGather. Plus a final dense layer with one output. 

I'm at best getting R2 test and validation values around 0.75, nowhere near the 0.9xx values reported in the paper. Tried 'index' and 'random' splits, and 100+ epochs instead of the recommended 20, to no avail.

Realise this is not necessarily an easy issue to pinpoint, but would be great if anyone had ideas what I'm missing to reproduce the MoleculeNet results.

**Comment by vid33** [comment]
Was the MoleculeNet stuff perhaps run prior to GraphConv being merged with TensorGraph, since the above variables, (n_filters, n_fully_connected_nodes, etc.) are as far as I can see not defined in GraphConvTensorGraph? To test the above hyperparameters, I actually changed values in the library graph_models.py file - so the origin of the problem may be related to a discrepancy in some detail here.

**Comment by miaecle** [comment]
@vid33 Yes, benchmarks are run on the previous graphconv model(see`examples/tox21/tox21_graph_conv.py` or `deepchem/molnet/run_benchmark_models.py`). New models are having identical structures so I don't think this will affect the score a lot. I have reproduced your result, which seems to be a tricky shifting issue.(actually the current version report the same score for graphconv on QM7) I suppose this is something due to the update of our data file(since QM7 did not come with connectivity and SMILES). For now, I recommend you to switch to deep tensor neural network(DTNN) to see if it can get a higher score.

**Comment by vid33** [comment]
Thanks for your help. It looks like you are right. The optimal hyperparameters do seem to reproduce things correctly for ESOL, which points to there being something wrong with the input data for QM7. Presumably, this will effect all models on QM7 then? 

Are you suggesting the qm7.csv file on deepchem s3 is corrupt somehow? That just contains SMILES and atomization energies.

FWIW, I'm actually specifically interested in benchmarking the graph conv stuff against some ideas I have, so am not so interested in other models. Would also prefer to use certain custom layers I've written as part of tensorgraph, and not have to go back and incorporate them into the older code.

If you can point me to something I can do, I'm happy to try and help with this, since I'd like to try and reproduce the MoleculeNet scores soon if possible.

**Comment by miaecle** [comment]
@vid33 Thanks for your interest! For QM7, as the original dataset only including 3D coordinates of atoms in the molecules(no explicit bonds), so models purely based on 3D coordinates should be fine, which include DTNN and tf_regression_ft(singletask network on coulomb matrix). Things could be broken when we try to generate SMILES from coordinates, this will affect all SMILES-based methods: ECFP-featurized methods, graphconv, weave, etc. So my suggest is to skip this dataset since we really don't recommend graphconv on it. A better alternative might be qm8, which comes with both 3D coordinates and SMILES.

**Comment by vid33** [comment]
Will try qm8, good suggestion. 

Agree that graphconv at least in the current implementation, is not optimal, and that 3D coordinate information should not be thrown away. It was partially for this reason that I wanted to reproduce the reported results, since they actually seem surprisingly good.
